### DISCUSSION TITLE
MOM6 Bug Fixes - Nov 2023 to Jul 2024

### DISCUSSION THREAD

**Discussion by yichengt900** [NOAA-GFDL/CEFI-regional-MOM6#76] [Announcements]
Hi CEFI-regional-MOM6 modelers,

Below is a table summarizing the new bug fix flags added from November 2023 to July 2024 (Thanks @theresa-morrison). Users of regional MOM6 configurations under development can use this information to avoid issues related to the identified bugs.

**Last Update: August 5, 2024 (WIP)**
| Flag | Default | Recommended | Pull Request | Notes |
|---------|---------|---------|---------|----------------------|
| MASS_WEIGHT_IN_PGF_NONBOUS_BUG| False| False| [PR 692](https://github.com/NOAA-GFDL/MOM6/pull/692)| Only used in non-Boussinesq cases with MASS_WEIGHT_IN_PRESSURE_GRADIENT set to true|
| FRICTWORK_BUG| True| False| [PR 663](https://github.com/NOAA-GFDL/MOM6/pull/663)| |
| [*]_USE_OM4_SUBCELLS| True| False| [PR 662](https://github.com/NOAA-GFDL/MOM6/pull/662)|There are many new parameters added with this suffix, all should be set to false|
| BACKSCATTER_UNDERBOUND| True| False| [PR 656](https://github.com/NOAA-GFDL/MOM6/pull/656)| Only applies when Lapalcian viscosity is negative (such as from backscatter)|
| VISC_REM_BUG| True| False| [PR 639](https://github.com/NOAA-GFDL/MOM6/pull/639)| This parameter controls the defaults of three individual flags, VISC_REM_TIMESTEP_FIX, VISC_REM_BT_WEIGHT_FIX, and VISC_REM_CONT_HVEL_FIX|
| MEKE_GM_SRC_ALT_SLOPE_BUG| True| False| [PR 620](https://github.com/NOAA-GFDL/MOM6/pull/620)| This parameter controls the defaults of three individual flags, VISC_REM_TIMESTEP_FIX, VISC_REM_BT_WEIGHT_FIX, and VISC_REM_CONT_HVEL_FIX|
| INIT_BOUNDARY_EXTRAP| False| True| [PR 608](https://github.com/NOAA-GFDL/MOM6/pull/608)| May only impact domains with ice cavities|
| HOR_DIFF_LIMIT_BUG|True| False| [PR 585](https://github.com/NOAA-GFDL/MOM6/pull/585)|If true and HOR_DIFF_ANSWER_DATE is 20240330 or below break symmetry |
| USE_CONT_THICKNESS| False| |[PR 579](https://github.com/NOAA-GFDL/MOM6/pull/579) | See also [PR 697](https://github.com/NOAA-GFDL/MOM6/pull/697) |
| TRIG_CHANNEL_DRAG_WIDTHS| True| | [PR 543](https://github.com/NOAA-GFDL/MOM6/pull/543)| SIf false, use a more accurate than the mathematically equivalent trigonometric version|
| UNSPLIT_DT_VISC_BUG| False| False| [PR536](https://github.com/NOAA-GFDL/MOM6/pull/536)| Replaces FIX_UNSPLIT_DT_VISC_BUG (was default true)|
| USTAR_GUSTLESS_BUG| False| False| [PR535](https://github.com/NOAA-GFDL/MOM6/pull/535)| Replaces FIX_USTAR_GUSTLESS_BUG (was default true)|
| USE_HUYNH_STENCIL_BUG| False| False| | |

**P.S**. In MOM6, “bugs” can refer to incorrect code or mathematically equivalent code that fails rigorous tests (such as rotational testing). Consequently, while these flags might alter the results of a simulation, they might not affect the solution at all. To verify whether solutions agree or disagree across tests, it is recommended to compare the `ocean.stats` files.

**Comment by yichengt900** [comment]
Contributed from @theresa-cordero :

Below are some new changes in MOM6 runtime parameters [since early September 2024](https://github.com/NOAA-CEFI-Regional-Ocean-Modeling/MOM6/commit/a6dd0fd0df55c9000b770b9e229d96485812280e):

New bug fix flags:
VISC_REM_BUG should be false
VISC_REM_TIMESTEP_BUG should be false
VISC_REM_BT_WEIGHT_BUG should be false
EPBL_MLD_ITER_BUG should be false

Other notable changes:
USE_POROUS_BARRIER  default was True, is now false
OBC_REMAPPING_SCHEME has been added, the default seems good
OBC_REMAPPING_USE_OM4_SUBCELLS default was True, is now False (*set in MOM_input)
DIAG_REMAPPING_USE_OM4_SUBCELLS default was True, is now false (*set in MOM_input)
SAL_USE_BPA is a new option that has been added, [see this PR](https://github.com/NOAA-GFDL/MOM6/pull/731), we may want this to be True, but it may be mainly for Boussinesq mode and seems to require more input files.
TIDE_*_PHASE_T0 default values have changed, I think related [to this PR](https://github.com/NOAA-GFDL/MOM6/pull/764) that changed the units of tidal frequencies
OBC_TIDE_ADD_NODE (and similar) became TIDE_ADD_NODE, old names will still work
MAX_RRHO_SALT_FINGERS default changed, not set in MOM_input. The new default preserves answers though, so you may not want to change. [PR 746](https://github.com/NOAA-GFDL/MOM6/pull/746)
DIRECT_EPBL_MIXING_CALC try with this set to true, it is "more efficient than the more general iterative approach" but will change answers. Part of a lot going on in [PR 717](https://github.com/NOAA-GFDL/MOM6/pull/717)

**Reply by amoebaliz** [reply]
My simulations crashed with DIRECT_EPBL_MIXING_CALC = True

I'm not 100% certain I adequately isolated that as the culprit but I have been running without it for now

**Comment by theresa-cordero** [comment]
Updating to commit [922d2a0](https://github.com/NOAA-CEFI-Regional-Ocean-Modeling/MOM6/commit/922d2a0a8d0fffed7907b670979e544c5be7e84b) April 2025:

**Bug fix**
`EPBL_MLD_ITER_BUG` fixes an "unlikely" bug when iterating to determine the mixed layer depth, that arises when the change between iterations exactly matches the tolerance. Should be set to `False` [PR 717](https://github.com/NOAA-GFDL/MOM6/pull/717)
`VISC_REM_TIMESTEP_FIX` has been renamed to `VISC_REM_TIMESTEP_BUG` which should be `False` [PR 745](https://github.com/NOAA-GFDL/MOM6/pull/745)
`VISC_REM_BT_WEIGHT_FIX` has been renamed to `VISC_REM_BT_WEIGHT_BUG` which should be `False` [PR 745](https://github.com/NOAA-GFDL/MOM6/pull/745)
`EXTERIOR_OBC_BUG` if true, recover a bug when OBCs are in the interior of the domain  [PR 814](https://github.com/NOAA-GFDL/MOM6/pull/814)
`RHO_PGF_REF_BUG` fixes a bug in Boussinesq finite volume pressure gradient force, regional models don't use Boussinesq mode so this can be left as the default true [PR 837](https://github.com/NOAA-GFDL/MOM6/pull/837)

**Default Change**
`SAL_SCALAR_VALUE` default was `0.01`, is now showing as `0.0` [PR 819](https://github.com/NOAA-GFDL/MOM6/pull/819)

**New Parameters**
`BACKSCAT_EBT_POWER` These parameters are related to the Backscatter code which is not used in the regional models. The defaults should cause no issue. [PR 706](https://github.com/NOAA-GFDL/MOM6/pull/706)

`MASS_WEIGHT_IN_PRESSURE_GRADIENT_TOP`, `CORRECTION_INTXPA`, `RESET_INTXPA_INTEGRAL`, `RESET_INTXPA_INTEGRAL_FLATTEST` and `MASS_WEIGHT_IN_PGF_VANISHED_ONLY` are related to correcting the pressure gradient force especially under ice sheets. See [PR 713](https://github.com/NOAA-GFDL/MOM6/pull/713), [PR 810](https://github.com/NOAA-GFDL/MOM6/pull/810) and [PR 843](https://github.com/NOAA-GFDL/MOM6/pull/843)

`SAL_USE_BPA` if true use bottom pressure anomaly (BPA) to calculate self-attraction and loading (SAL). if true a long term mean reference bottom pressure is required. [PR 731](https://github.com/NOAA-GFDL/MOM6/pull/731)
`BOUSSINESQ_SAL_TIDES` Applies to Boussinesq mode, which is not used by Regional models at this time.  [PR 731](https://github.com/NOAA-GFDL/MOM6/pull/731)

`BS_USE_SQG_STRUCT`  sets the vertical structure of backscatter [PR 738](https://github.com/NOAA-GFDL/MOM6/pull/738)
`SQG_EXPO` changes the shape of the surface quasigeostrophy (SQG) vertical structure. [PR 738](https://github.com/NOAA-GFDL/MOM6/pull/738)
`KHTH_USE_SQG_STRUCT` sets the vertical structure of `KHTH` [PR 738](https://github.com/NOAA-GFDL/MOM6/pull/738)
`KHTR_USE_SQG_STRUCT` sets the vertical structure of `KHTR` [PR 738](https://github.com/NOAA-GFDL/MOM6/pull/738)
`KD_GL90_USE_SQG_STRUCT ` sets the vertical structure of  `KD` [PR 738](https://github.com/NOAA-GFDL/MOM6/pull/738)

`DT_TRACER_ADVECT` and `TRADV_SPANS_COUPLING`, default to equal `DT_THERM` and `THERMO_SPANS_COUPLING`. These were added to test the sensitivity of the model to the tracer advection timestep. I recommend the defaults for now. [PR 757](https://github.com/NOAA-GFDL/MOM6/pull/757)

`USE_FILTER` if set to true enables the use of band-pass filters to appply wave drag to tidal velocities. 
[PR 772](https://github.com/NOAA-GFDL/MOM6/pull/772) and [PR 815](https://github.com/NOAA-GFDL/MOM6/pull/815)
`BT_LINEAR_FREQ_DRAG` If true, apply frequency-dependent drag to the tidal velocities when `USE_FILTER` is true,  [PR 815](https://github.com/NOAA-GFDL/MOM6/pull/815)

`OPACITY_VALUES_MANIZZA` This PR was a refactoring of `MOM_opacity` which introduced these parameters. The defaults restore old solutions. I don't recommend these be changed unless you are testing a new opacity scheme. [PR 775](https://github.com/NOAA-GFDL/MOM6/pull/775)
`CHOROPHYLL_POWER_MANIZZA` also in [PR 775](https://github.com/NOAA-GFDL/MOM6/pull/775)
`SW_PEN_FRAC_COEFS_MOREL` also in [PR 775](https://github.com/NOAA-GFDL/MOM6/pull/775)
`OPACITY_BAND_WAVELENGTHS` also in [PR 775](https://github.com/NOAA-GFDL/MOM6/pull/775)

`RESTART_SYMMETRIC_CHECKSUMS` These flags are useful for testing grid rotation across restarts. For production runs they can be left as the default.  [PR 787](https://github.com/NOAA-GFDL/MOM6/pull/787)
`RESTART_UNSIGNED_ZEROS` [PR 787](https://github.com/NOAA-GFDL/MOM6/pull/787)

`SET_DTBT_USE_BT_CONT` This is most useful for the Great Lakes model, can be kept at the default of true.  [PR 823](https://github.com/NOAA-GFDL/MOM6/pull/823)

`VERTEX_SHEAR_THICKNESS_MEAN` if true, apply thickness weighting when averaging to vertex point. I think most regional models do not use vertex shear, so this may not apply. 
[PR 846](https://github.com/NOAA-GFDL/MOM6/pull/846)
`VERTEX_SHEAR_GEOMETRIC_MEAN` default is false, consider testing with true. 
[PR 846](https://github.com/NOAA-GFDL/MOM6/pull/846)

**Reply by yichengt900** [reply]
@theresa-cordero, Thanks for the updates!
